### PR TITLE
refactor(datastore): Added delete by id option that takes ModelSchema

### PR DESCRIPTION
**Pull request by royjit**
*Description of changes:*
- Added a new escape hatch api for delete by id
- Fixed graphql request decorator while using a dependent model.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Auth/AWSCognitoAuthPlugin/ClientBehavior/AWSCognitoAuthPlugin+ClientBehavior.swift [hunk 1]
```diff
@@ -83,6 +83,8 @@ extension AWSCognitoAuthPlugin {
                                 options: AuthSocialWebUISignInOperation.Request.Options?,
                                 listener: AuthSocialWebUISignInOperation.ResultListener?)
         -> AuthSocialWebUISignInOperation {
+        
+
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSCognitoAuthPlugin/ClientBehavior/AWSCognitoAuthPlugin+ClientBehavior.swift:87]: nit: vertical whitespace. Please confirm that SwiftFormat/Lint would have caught this. If not, we need to figure out why--either the tool isn't configured right or it's not being invoked on build?
```suggestion
```

File: AmplifyPlugins/Core/AWSPluginsCore/Model/Support/Model+GraphQL.swift [hunk 2]
```diff
@@ -44,7 +44,16 @@ extension Model {
                 if case let .belongsTo(_, targetName) = field.association {
                     fieldName = targetName ?? fieldName
                 }
-                input[fieldName] = (value as? Model)?.id
+                if let modelValue = value as? Model {
+                    input[fieldName] = modelValue.id
+
+                } else if let value = value as? [String: JSONValue],
+                   case let .string(primaryKeyValue) = value[modelSchema.primaryKey.name] {
```
Comment (palpatim) [inline AmplifyPlugins/Core/AWSPluginsCore/Model/Support/Model+GraphQL.swift:51]: nit: prefer `let` and `var` bindings for each individual element. Turns out it's not just a clearer style practice, it also avoids some unexpected behavior if your bound variable shadows something in an outer scope: https://google.github.io/swift/#pattern-matching
```suggestion
                   case .string(let primaryKeyValue) = value[modelSchema.primaryKey.name] {
```

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/AWSDataStorePlugin+DataStoreBaseBehavior.swift [hunk 3]
```diff
@@ -111,8 +111,15 @@ extension AWSDataStorePlugin: DataStoreBaseBehavior {
...
                                  completion: @escaping DataStoreCallback<Void>) {
+        delete(modelType, modelSchema: modelType.schema, withId: id, completion: completion)
+    }
+    
+    public func delete<M: Model>(_ modelType: M.Type,
+                                 modelSchema: ModelSchema,
+                                 withId id: String,
+                                 completion: @escaping DataStoreCallback<Void>) {
         reinitStorageEngineIfNeeded()
-        storageEngine.delete(modelType, modelSchema: modelType.schema, withId: id) { result in
```
Comment (palpatim) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/AWSDataStorePlugin+DataStoreBaseBehavior.swift:115]: Have you tried a global text search on `modelType.schema` to see if there are other places this needs to be updated?

Comment (royjit) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/AWSDataStorePlugin+DataStoreBaseBehavior.swift:115]: Yes, most of them are covered in this PR. There are 2-3 places related to subscription and usage of untyped model, which I will address in another PR.

File: AmplifyPlugins/Auth/AWSCognitoAuthPlugin/ClientBehavior/AWSCognitoAuthPlugin+ClientBehavior.swift (hunk 1 above)
Comment (royjit) [reply AmplifyPlugins/Auth/AWSCognitoAuthPlugin/ClientBehavior/AWSCognitoAuthPlugin+ClientBehavior.swift]: This is from Auth plugin which was not caught while building datastore plugin. Swift format/lint is working as expected.

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift [hunk 4]
```diff
@@ -252,8 +252,8 @@ final class StorageEngine: StorageEngineBehavior {
         let transactionResult = queryAndDeleteTransaction(modelType, modelSchema: modelSchema, predicate: predicate)
 
         guard modelSchema.isSyncable, let syncEngine = self.syncEngine else {
-            if !modelType.schema.isSystem {
-                log.error("Unable to sync modelType (\(modelType)) where isSyncable is false")
+            if !modelSchema.isSystem {
+                log.error("Unable to sync mdoel (\(modelSchema.name)) where isSyncable is false")
```
Comment (palpatim) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift:256]: nit: spelling
```suggestion
                log.error("Unable to sync model (\(modelSchema.name)) where isSyncable is false")
```

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/MutationSync/OutgoingMutationQueue/ProcessMutationErrorFromCloudOperation.swift [hunk 5]
```diff
@@ -300,7 +300,13 @@ class ProcessMutationErrorFromCloudOperation: AsynchronousOperation {
             return
         }
 
-        storageAdapter.delete(untypedModelType: modelType, modelSchema: modelType.schema, withId: id) { response in
+        guard let modelSchema = ModelRegistry.modelSchema(from: modelName) else {
+            let error = DataStoreError.unknown("Invalid Model \(modelName)", "")
```
Comment (palpatim) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/MutationSync/OutgoingMutationQueue/ProcessMutationErrorFromCloudOperation.swift:304]: Recovery suggestion?

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/SubscriptionSync/ReconcileAndLocalSave/ReconcileAndLocalSaveOperation.swift [hunk 6]
```diff
@@ -214,8 +214,17 @@ class ReconcileAndLocalSaveOperation: AsynchronousOperation {
 
     private func saveDeleteMutation(storageAdapter: StorageEngineAdapter, remoteModel: RemoteModel) {
         log.verbose(#function)
-        guard let modelType = ModelRegistry.modelType(from: remoteModel.model.modelName) else {
-            let error = DataStoreError.invalidModelName(remoteModel.model.modelName)
+        let modelInstance = remoteModel.model.instance
+        let modelName: String
+        if let jsonModel = modelInstance as? JSONValueHolder,
+           let modelNameFromJson = jsonModel.jsonValue(for: "__typename") as? String {
```
Comment (palpatim) [inline AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/SubscriptionSync/ReconcileAndLocalSave/ReconcileAndLocalSaveOperation.swift:220]: Do we have the `__typename` field modeled as a protocol or typename somewhere? It feels a little weird to suddenly be diving into conflict resolution payload field names here

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift (hunk 4 above)
Comment (royjit) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Storage/StorageEngine.swift:256]: Fixed

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/MutationSync/OutgoingMutationQueue/ProcessMutationErrorFromCloudOperation.swift (hunk 5 above)
Comment (royjit) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/MutationSync/OutgoingMutationQueue/ProcessMutationErrorFromCloudOperation.swift:304]: Changed to invalidModelName error, it has its own recovery suggesiton.

File: AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/SubscriptionSync/ReconcileAndLocalSave/ReconcileAndLocalSaveOperation.swift (hunk 6 above)
Comment (royjit) [reply AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/SubscriptionSync/ReconcileAndLocalSave/ReconcileAndLocalSaveOperation.swift]: Removed this line as per our discussion and uses ModelSchema directly. I will see if I can do the same in another pr for the other usage of `__typename`.

**Review by palpatim — APPROVED** [review]

**Pull request merged**